build/Build.Steps.cs

```diff
@@ -66,15 +66,43 @@ partial class Build
         .Unlisted()
         .Executes(() => ControlFlow.ExecuteWithRetry(() =>
         {
+            var projectsToRestore = Solution.GetCrossPlatformManagedProjects();
+
             if (IsWin)
             {
-                DotNetRestore(s => s
-                    .SetProjectFile(Solution)
-                    .SetVerbosity(DotNetVerbosity.Normal)
-                    .SetProperty("configuration", BuildConfiguration.ToString())
-                    .When(!string.IsNullOrEmpty(NugetPackageDirectory), o =>
-                        o.SetPackageDirectory(NugetPackageDirectory)));
+                projectsToRestore = projectsToRestore.Concat(Solution.GetWindowsOnlyTestApplications());
+            }
 
+            foreach (var project in projectsToRestore)
+            {
+                if (TestPackageVersions.Versions.TryGetValue(project.Name, out var testedPackageVersions))
+                {
+                    foreach (var testedPackageVersion in testedPackageVersions)
+                    {
+                        DotNetRestore(s => s
+                            .SetProjectFile(project)
+                            .SetVerbosity(DotNetVerbosity.Normal)
+                            .SetProperty("configuration", BuildConfiguration.ToString())
+                            .SetProperty("TestedPackageVersion", testedPackageVersion)
+                            .SetPlatform(Platform)
+                            .When(!string.IsNullOrEmpty(NugetPackageDirectory), o =>
+                                o.SetPackageDirectory(NugetPackageDirectory)));
+                    }
+                }
+                else
+                {
+                    DotNetRestore(s => s
+                            .SetProjectFile(project)
+                            .SetVerbosity(DotNetVerbosity.Normal)
+                            .SetProperty("configuration", BuildConfiguration.ToString())
+                            .SetPlatform(Platform)
+                            .When(!string.IsNullOrEmpty(NugetPackageDirectory), o =>
```

Comment:
I _think_ that this could be done with a `.CombineWith` to avoid the `if` statement and the almost identical branches.

Reply:
I mean not this specific line but the `DotNetRestore` itself.

Reply:
I do not find the way to fully remove branching (When is not working well with CobineWith - due to possible return type differences), but I some improvements was possible. ba965160f94c1aad4e45de4f84e3987069e0d399